mpfr: speed improvements

vtjnash opened the pull request:
Unlike GMP / BigInt, the MPFR / BigFloat library exposes an interface for external allocation:
http://www.mpfr.org/mpfr-current/mpfr.html#Custom-Interface

Using that allows us to use our fast memory-pool gc, and avoid adding finalizers and use the slow malloc/free functions, and allows us to better track memory pressure.

```julia
function benchwork(x, y)
    for i in 1:100000000
          x += y
     end
     return x
end
@time benchwork(big"0.0", big"23.0")
```
```
 12.923691 seconds (200.00 M allocations: 8.196 GiB, 31.56% gc time)
  5.469907 seconds (200.00 M allocations: 10.431 GiB, 6.15% gc time)
```
(`@profile` now indicates that this is now about equal parts mpfr time and allocation-related time)

vtjnash committed:
mpfr: speed improvements

Unlike GMP / BigInt, the MPFR library exposes an interface for external allocation:
http://www.mpfr.org/mpfr-current/mpfr.html#Custom-Interface

Using that allows us to use our fast memory-pool gc
and avoid adding finalizers and use the slow malloc/free functions.
(and in some cases, some of it might even end up on the stack!)

Also switch to allocating lgamma_sign on the stack, for better performance (and thread-safety)

Also fixes Serialization to preserve BigFloat precision.

StefanKarpinski commented:
Does this help with serializing and deserializing BigFloats?

Keno commented on base/mpfr.jl:
```diff
@@ -55,18 +56,29 @@ mutable struct BigFloat <: AbstractFloat
     exp::Clong
     d::Ptr{Limb}
+    # _d::Buffer{Limb} # Julia gc handle for memory @ d
+    _d::String # Julia gc handle for memory @ d (optimized)
```
Mutating immutable memory is dangerous. If you must have this add a `FixedSizeBuffer` type.

vtjnash replied:
It's not _that_ dangerous, it's only bad if this pointer gets leaked anywhere else, but it's purely internal. I mean, it's not like there's a `deepcopy` method definition here that assumes they can't be mutated or anything like that. No, that would be crazy talk, and definitely not a bug in this very PR. :P

Keno replied:
Immutable memory is still being overwritten (since you construct it and then overwrite it with the result of an operation). I agree that it's much less likely to be harmful, since the compiler shouldn't be able to know the objectid of a newly-allocated bigfloat that gets subsequently modified (as long as the BigFloat itself isn't mutated more than once), but it does sound fairly risky.

vtjnash replied:
Isn't that just the same concern as any other usage of String?

vtjnash committed:
Serialization: fix a handful of implicit return mistakes

This would have caused us to perform a number of unncessary allocations and method specializations

vtjnash commented:
> Does this help with serializing and deserializing BigFloats?

Forgot to add that to the initial commit – now it does :)

KristofferC commented:
@nanosoldier `runbenchmarks("scalar", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27331#issuecomment-396489395) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2c7d22b7b49d3aee62d8862624a40b8cc5e095d9/dd80ed0_vs_83ce7ba/report.md). cc @ararslan

KristofferC commented:
Nice :)

vtjnash commented:
So, about 25% faster. (The allocation difference is just previous measurement error)

Sacha0 commented:
> Nice :)

An understatement! That's fantastic! 🎉

KristofferC commented:
It was lucky that it had memory regressions or we wouldn't have seen the speed improvements because the threshold is put to 50% (e.g. needs to become twice as fast for it to be considered insignificant). Feels kinda high...

dlfivefifty commented:
FYI this PR is probably the cause of a SEGFAULT in ` gamma(big(0)) ` :

https://github.com/JuliaMath/SpecialFunctions.jl/issues/101#issuecomment-399709741
